keras_nlp/applications/bert.py

```diff
@@ -66,118 +67,118 @@ class BertModel(keras.Model):
-        # these weights for fine-tuning tasks.
-        self._pooler_layer = keras.layers.Dense(
-            units=hidden_size,
-            activation="tanh",
-            kernel_initializer=self.initializer,
-            name="pooler_dense",
-        )
+            )(x, padding_mask=input_mask)
 
-        self.inputs = dict(
-            input_ids=keras.Input(shape=(None,), dtype=tf.int32),
-            input_mask=keras.Input(shape=(None,), dtype=tf.int32),
-            segment_ids=keras.Input(shape=(None,), dtype=tf.int32),
-        )
-
-    def call(self, inputs):
-        if isinstance(inputs, dict):
-            input_ids = inputs.get("input_ids")
-            input_mask = inputs.get("input_mask")
-            segment_ids = inputs.get("segment_ids")
-        else:
-            raise ValueError(f"Inputs should be a dict. Received: {inputs}.")
-
-        word_embeddings = None
-        word_embeddings = self._embedding_layer(input_ids)
-        position_embeddings = self._position_embedding_layer(word_embeddings)
-        type_embeddings = self._type_embedding_layer(segment_ids)
-
-        embeddings = word_embeddings + position_embeddings + type_embeddings
-        embeddings = self._embedding_norm_layer(embeddings)
-        embeddings = self._embedding_dropout(embeddings)
-
-        x = embeddings
-        for layer in self._transformer_layers:
-            x = layer(x, padding_mask=input_mask)
+        # Construct the two BERT outputs, and apply a dense to the pooled output.
         sequence_output = x
-        pooled_output = self._pooler_layer(x[:, 0, :])  # 0 is the [CLS] token.
-        return sequence_output, pooled_output
+        pooled_output = keras.layers.Dense(
+            hidden_size,
+            activation="tanh",
+            name="pooled_dense",
+        )(x[:, CLS_INDEX, :])
+
+        # Instantiate using Functional API Model constructor
+        super(Bert, self).__init__(
+            inputs={
+                "input_ids": token_id_input,
+                "segment_ids": segment_id_input,
+                "input_mask": input_mask,
+            },
+            # TODO(jbischof): Consider list output
+            outputs={
+                "sequence_output": sequence_output,
+                "pooled_output": pooled_output,
+            },
+            **kwargs
+        )
+        # All references to `self` below this line
+        self.inner_activation_fn = inner_activation_fn
+        self.initializer_fn = initializer_fn
+        self.vocab_size = vocab_size
+        self.hidden_size = hidden_size
+        self.num_layers = num_layers
+        self.num_attention_heads = num_attention_heads
+        self.max_sequence_length = max_sequence_length
+        self.type_vocab_size = type_vocab_size
+        self.inner_size = inner_size
+        self.inner_activation = keras.activations.get(inner_activation)
+        self.initializer_range = initializer_range
+        self.dropout = dropout
 
     def get_embedding_table(self):
```

Comment:
This was inherited from model garden essentially, I'm not sure we want it. Should we drop it for now?

Maybe `model.token_embedding.embeddings` would be better?

Reply:
That doesn't seem to work with error: `AttributeError: 'KerasTensor' object has no attribute 'embeddings'`. It seems like we need to grab the layer, right?

Reply:
Yea, we would have to add the properties to the class ourselves, in init, after chaining up to super.
```
super().__init__(inputs=, outputs=)
self.token_embedding = token_embedding
self.position_embedding = position_embedding
self.segment_embedding = segment_embedding
```

That me be more robust that looking up by name, depending on the get_uid naming thread above. And feels a little more general purpose to allow access to all the embeddings.

Reply:
Actually I did this before getting that error I pasted above. It seems that was is assigned is a `KerasTensor` and not a `Layer`

Reply:
oh, yeah, you would have to update your code above to have the layer as a local variable

```
# Embed tokens, positions, and segment ids.
token_embedding = keras.layers.Embedding(
    input_dim=vocab_size,
    output_dim=hidden_size,
    name=TOKEN_EMBEDDING_LAYER_NAME,
)
embedded_tokens = token_embedding(token_id_input)
position_embedding = PositionEmbedding(
    initializer=initializer_fn,
    sequence_length=max_sequence_length,
    name="position_embedding",
)
embedded_positions = position_embedding(embedded_tokens)
...
super().__init__(...)
...
self.token_embedding = token_embedding
self.position_embedding = position_embedding
```

Reply:
Thanks, this makes sense now! The functional API is very cool but not much like generic OOP.